src/function/aggregate/sorted_aggregate_function.cpp

```diff
@@ -58,26 +46,102 @@ struct SortedAggregateBindData : public FunctionData {
-		       sort_types == other.sort_types;
+		if (function != other.function) {
+			return false;
+		}
+		if (orders.size() != other.orders.size()) {
+			return false;
+		}
+		for (size_t i = 0; i < orders.size(); ++i) {
+			if (!orders[i].Equals(other.orders[i])) {
+				return false;
+			}
+		}
+		return true;
 	}
 
+	BufferManager &buffer_manager;
 	AggregateFunction function;
 	vector<LogicalType> arg_types;
 	unique_ptr<FunctionData> bind_info;
-
-	vector<OrderType> order_sense;
-	vector<OrderByNullType> null_order;
 	vector<LogicalType> sort_types;
+	vector<BoundOrderByNode> orders;
 };
 
 struct SortedAggregateState {
-	SortedAggregateState()
-	    : arguments(Allocator::DefaultAllocator()), ordering(Allocator::DefaultAllocator()), nsel(0) {
+	static const idx_t BUFFER_CAPACITY = STANDARD_VECTOR_SIZE;
+
+	SortedAggregateState() : nsel(0) {
 	}
 
-	ChunkCollection arguments;
-	ChunkCollection ordering;
+	unique_ptr<ColumnDataCollection> arguments;
+	unique_ptr<ColumnDataCollection> ordering;
+
+	DataChunk sort_buffer;
+	DataChunk arg_buffer;
+
+	static inline void InitializeBuffer(DataChunk &chunk, const vector<LogicalType> &types) {
+		if (!chunk.ColumnCount() && !types.empty()) {
+			chunk.Initialize(Allocator::DefaultAllocator(), types);
+		}
+	}
+
+	void Flush(SortedAggregateBindData &order_bind) {
+		if (ordering) {
+			return;
+		}
+
+		ordering = make_unique<ColumnDataCollection>(order_bind.buffer_manager, order_bind.sort_types);
+		InitializeBuffer(sort_buffer, order_bind.sort_types);
+		ordering->Append(sort_buffer);
+
+		arguments = make_unique<ColumnDataCollection>(order_bind.buffer_manager, order_bind.arg_types);
+		InitializeBuffer(arg_buffer, order_bind.arg_types);
+		arguments->Append(arg_buffer);
+	}
+
+	void Update(SortedAggregateBindData &order_bind, DataChunk &sort_chunk, DataChunk &arg_chunk) {
+		// Lazy instantiation of the buffer chunks
+		InitializeBuffer(sort_buffer, order_bind.sort_types);
+		InitializeBuffer(arg_buffer, order_bind.arg_types);
+
+		if (sort_chunk.size() + sort_buffer.size() > BUFFER_CAPACITY) {
+			Flush(order_bind);
+		}
+		if (ordering) {
+			ordering->Append(sort_chunk);
+			arguments->Append(arg_chunk);
+		} else {
+			sort_buffer.Append(sort_chunk, true);
+			arg_buffer.Append(arg_chunk, true);
+		}
+	}
+
+	void Combine(SortedAggregateBindData &order_bind, SortedAggregateState &other) {
+		if (other.ordering) {
+			// Force CDC if the other hash it
+			Flush(order_bind);
+			ordering->Combine(*other.ordering);
+			arguments->Combine(*other.arguments);
+		} else if (other.sort_buffer.size()) {
+			Update(order_bind, other.sort_buffer, other.arg_buffer);
+		}
+	}
+
+	void Finalize(LocalSortState &local_sort) {
+		if (ordering) {
+			const auto chunk_count = ordering->ChunkCount();
+			for (idx_t chunk_idx = 0; chunk_idx < chunk_count; ++chunk_idx) {
+				ordering->FetchChunk(chunk_idx, sort_buffer);
+				arguments->FetchChunk(chunk_idx, arg_buffer);
```

Comment:
`FetchChunk` is cannot do zero-copy reads from the CDC, and is therefore much less efficient than scanning the CDC using `ColumnDataCollection::Scan(ColumnDataScanState &state, DataChunk &result)`. Doing the zero-copy read here is fine since the data is copied to the sort state when you sink it anyway.

Reply:
Thanks - this is why I dragged you in!

Reply:
Ah when I do that the sort code crashes with a bad memory access.

Reply:
My bad - I wasn't Resetting the DataChunks.